cpp/roaring64map.hh

```diff
@@ -0,0 +1,796 @@
+                                      std::numeric_limits<uint32_t>::max());
+            roarings[start_high].setCopyOnWrite(copyOnWrite);
+        }
+
+        roarings[start_high].flip(std::numeric_limits<uint32_t>::min(), end_low);
+        roarings[start_high].setCopyOnWrite(copyOnWrite);
+    }
+
+    /**
+     *  Remove run-length encoding even when it is more space efficient
+     *  return whether a change was applied
+     */
+    bool removeRunCompression() {
+        return std::accumulate(roarings.begin(), roarings.end(), false,
+            [](bool previous, std::pair<const uint32_t, Roaring>& map_entry) {
+                return map_entry.second.removeRunCompression() && previous;
+            });
+    }
+
+    /** convert array and bitmap containers to run containers when it is more
+     * efficient;
+     * also convert from run containers when more space efficient.  Returns
+     * true if the result has at least one run container.
+     * Additional savings might be possible by calling shrinkToFit().
+     */
+    bool runOptimize() {
+        return std::accumulate(roarings.begin(), roarings.end(), false,
+            [](bool previous, std::pair<const uint32_t, Roaring>& map_entry) {
+                return map_entry.second.runOptimize() && previous;
+            });
+    }
+
+    /**
+     * If needed, reallocate memory to shrink the memory usage. Returns
+     * the number of bytes saved.
+    */
+    size_t shrinkToFit() {
+        size_t savedBytes = 0;
+        auto iter = roarings.begin();
+        while (iter != roarings.cend()) {
+            if (iter->second.isEmpty()) {
+                // empty Roarings are 84 bytes
+                savedBytes += 88;
+                roarings.erase(iter++);
+            } else {
+                savedBytes += iter->second.shrinkToFit();
+                iter++;
+            }
+        }
+        return savedBytes;
+    }
+
+    /**
+     * Iterate over the bitmap elements. The function iterator is called once
+     * for
+     *  all the values with ptr (can be NULL) as the second parameter of each
+     * call.
+     *
+     *  roaring_iterator64 is simply a pointer to a function that returns void,
+     *  and takes (uint64_t,void*) as inputs.
+     */
+    void iterate(roaring_iterator64 iterator, void *ptr) const {
+        std::for_each(roarings.begin(), roarings.cend(),
+            [=](const std::pair<uint32_t, Roaring>& map_entry) {
+                roaring_iterate64(map_entry.second.roaring, iterator, uint64_t(map_entry.first) << 32, ptr);
+            });
+    }
+
+    /**
+     * If the size of the roaring bitmap is strictly greater than rank, then
+     this
+       function returns true and set element to the element of given rank.
+       Otherwise, it returns false.
+     */
+    bool select(uint64_t rank, uint32_t *element) const {
+        for (const auto& map_entry : roarings) {
+            uint64_t sub_cardinality = (uint32_t)map_entry.second.cardinality();
```

Comment:
this was supposed to be a (uint64_t) cast to keep Visual Studio from complaining